### PR TITLE
add structured config and unit tests

### PR DESCRIPTION
**Pull request by AmberCharitos**
Adds structured config to standardise the validation process and have all config values present with default value of `None`. Also means that all charm functionality will halt on a bad config - not allowing any further events except config changed.

I also resolved an error in a docstring regarding the restart action.

### CODE REVIEW

File: tests/unit/test_structured_config.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    check_valid_values(_harness, "charm-function", accepted_values)
+
+    # self-registration-role
+    check_invalid_values(_harness, "self-registration-role", erroneus_values)
+    accepted_values = ["Admin", "Alpha", "Gamma", "Public", "sql_lab"]
+    check_valid_values(_harness, "self-registration-role", accepted_values)
+
+
+def check_valid_values(_harness, field: str, accepted_values: list) -> None:
+    """Check the correcteness of the passed values for a field.
```
Comment (kelkawi-a) [inline tests/unit/test_structured_config.py:55]: `correctness`*

**Review by kelkawi-a — APPROVED** [review]
LGTM, neat work!

File: src/charm.py [hunk 2]
```diff
@@ -198,43 +197,21 @@ def _on_restart(self, event):
...
+        if "superset-secret-key" in self.config:
+            superset_secret = self.config["superset-secret-key"]
+        else:
+            superset_secret = generate_random_string(32)
+
         env = {
             "SUPERSET_SECRET_KEY": superset_secret,
             "ADMIN_PASSWORD": self.config["admin-password"],
-            "CHARM_FUNCTION": charm_function,
+            "CHARM_FUNCTION": self.config["charm-function"].value,
```
[the remark this answers is not in the document]
Comment (AmberCharitos) [reply src/charm.py:214]: Yes exactly, it's to access the string from the Enum.

File: tests/unit/test_structured_config.py (hunk 1 above)
Comment (AmberCharitos) [reply tests/unit/test_structured_config.py:55]: Ah thank you!

**Pull request merged**